dig.go

```diff
@@ -492,21 +492,37 @@ func cycleError(cycle []key, last key) error {
 	return errors.New(b.String())
 }
 
-func detectCycles(n *node, graph map[key]*node, path []key) error {
-	for _, p := range path {
-		if p == n.key {
-			return cycleError(path, n.key)
+func detectCycles(param param, graph map[key]*node, path []key) error {
+	switch p := param.(type) {
+	case paramList:
+		for _, p := range p.Params {
+			if err := detectCycles(p, graph, path); err != nil {
+				return err
+			}
 		}
-	}
-	path = append(path, n.key)
-	for _, dep := range n.Params.Dependencies() {
-		depNode, ok := graph[dep.key]
+	case paramSingle:
+		k := key{name: p.Name, t: p.Type}
+		for _, p := range path {
+			if p == k {
+				return cycleError(path, k)
+			}
+		}
+		path = append(path, k)
+
+		n, ok := graph[k]
 		if !ok {
-			continue
+			return nil
 		}
-		if err := detectCycles(depNode, graph, path); err != nil {
-			return err
+
+		return detectCycles(n.Params, graph, path)
+	case paramObject:
+		for _, f := range p.Fields {
+			if err := detectCycles(f.Param, graph, path); err != nil {
+				return err
+			}
 		}
+	default:
+		panic(fmt.Sprintf("unknown param type %T", param))
```

Comment:
This is one of those times when I really miss sum types and compiler-enforced exhaustiveness when dealing with them :/

Can we make this error message clearly indicate that the user can't fix this panic? It's an error in dig, and the user should immediately file an issue.

```
panic(fmt.Sprintf("Sorry - you've found a bug in dig! Please file an issue at https://github.com/uber-go/dig/issues/. Details: unknown param type %T", param))
```